conbench/app/_plots.py

```diff
@@ -186,41 +187,67 @@ def _source(
+    # finish) time of the corresponding benchmark case run (the `timestamp`
+    # property on the `BenchmarkCreate` schema in the Conbench API). Are these
+    # tz-aware or tz-naive but in UTC?
+    datetimes = [dateutil.parser.isoparse(x["timestamp"]) for x in data]
+    date_strings = [d.strftime("%Y-%m-%d %H:%M %Z") for d in datetimes]
 
-    points, means = [], []
+    points, values_with_unit = [], []
     if alert_min:
         for x in data:
             alert = 5 * float(x["distribution_stdev"])
             points.append(float(x["distribution_mean"]) - alert)
-            means.append(unit_fmt(points[-1], unit))
+            values_with_unit.append(unit_fmt(points[-1], unit))
     elif alert_max:
         for x in data:
             alert = 5 * float(x["distribution_stdev"])
             points.append(float(x["distribution_mean"]) + alert)
-            means.append(unit_fmt(points[-1], unit))
+            values_with_unit.append(unit_fmt(points[-1], unit))
     else:
         points = [x[key] for x in data]
         # Note(JP): If `means` is just string-formatted `points` then this is
         # kind of acknowledging that `points` is always a collection of mean
         # values. It seems that this is a string value field that is only
         # ever used for tooltip generation?
-        means = [unit_fmt(float(x[key]), unit) for x in data if x[key]]
+        values_with_unit = [unit_fmt(float(x[key]), unit) for x in data if x[key]]
 
     if formatted:
         # Note(JP): why would we want to calculate the raw numeric data from
         # the tooltip string labels?
-        points = [x.split(" ")[0] for x in means]
-
-    return bokeh.models.ColumnDataSource(
-        data={
-            "x": dates,
-            "y": points,
-            "commit_messages": commit_messages,
-            "commit_hashes_short": ["#" + d["sha"][:7] for d in data],
-            "means": means,
-        }
-    )
+        points = [x.split(" ")[0] for x in values_with_unit]
+
+    dsdict = {
+        "x": datetimes,
+        # Note(JP): maybe rename `points` into `y_values_for_plotting`
+        "y": points,
+        # List of human-readable date strings, corresponding to
+        # the `datetimes` objects for the time axis
+        "date_strings": date_strings,
+        # List of (truncated) commit messages.
+        "commit_messages": commit_messages,
+        # List of short commit hashes with hashtag prefix
+        "commit_hashes_short": ["#" + d["sha"][:8] for d in data],
+        "relative_benchmark_urls": [f'/benchmarks/{d["benchmark_id"]}' for d in data],
+        # Stringified values (truncated, with unit)
+        "values_with_unit": values_with_unit,
+    }
+
+    multisample, multisample_count = _inspect_for_multisample(data)
+    if multisample:
+        # Add a column to the ColumnarDataSource: stringified individual
+        # samples (with units).
+        strings = []
+        for d in data:
+            samples = d["data"]
```

Comment:
There is an example for where type annotations will help so so so soooooooooo much. Without type annotations one has to keep a lot of mental overhead about what `data` and `data["data"]` even are. With type annotations and proper editor setup a mouse-over hover effect simply provides the answer.